dashboard/dashboard.go

```diff
@@ -0,0 +1,314 @@
+	"fmt"
+	"github.com/ethereum/go-ethereum/log"
+	"github.com/ethereum/go-ethereum/p2p"
+	"github.com/ethereum/go-ethereum/rpc"
+	"github.com/rcrowley/go-metrics"
+	"golang.org/x/net/websocket"
+	"io/ioutil"
+	"net"
+	"net/http"
+	"sync"
+	"sync/atomic"
+	"time"
+)
+
+const (
+	processorSampleLimit = 200
+	memorySampleLimit    = 200
+	trafficSampleLimit   = 200
+)
+
+var (
+	nextId uint32 // Next connection id
+)
+
+type dashboard struct {
+	config *Config
+
+	listener net.Listener
+	conns    map[uint32]*client // Currently live websocket connections
+	Metrics  *metricSamples     `json:"metrics,omitempty"`
+	Stats    *status            `json:"stats,omitempty"`
+	lock     sync.RWMutex       // Lock protecting the dashboard's internals
+
+	closing chan chan error // Channel used for graceful exit
+	wg      sync.WaitGroup
+}
+
+type client struct {
+	conn   *websocket.Conn              // Particular live websocket connection
+	msg    chan *map[string]interface{} // Message queue for the update messages
+	logger log.Logger                   // Logger for the particular live websocket connection
+}
+
+type metricSamples struct {
+	Processor []*data `json:"processor,omitempty"`
+	Memory    []*data `json:"memory,omitempty"`
+}
+
+type data struct {
+	Time  time.Time `json:"time,omitempty"`
+	Value float64   `json:"value,omitempty"`
+}
+
+type status struct {
+	Peers int `json:"peers,omitempty"`
+	Block int `json:"block,omitempty"`
+}
+
+// New creates a new dashboard instance with the given configuration.
+func New(config *Config) (*dashboard, error) {
+	return &dashboard{
+		conns:   make(map[uint32]*client),
+		config:  config,
+		Metrics: &metricSamples{},
+		closing: make(chan chan error),
+	}, nil
+}
+
+// Protocols is a meaningless implementation of node.Service.
+func (db *dashboard) Protocols() []p2p.Protocol { return nil }
+
+// APIs is a meaningless implementation of node.Service.
+func (db *dashboard) APIs() []rpc.API { return nil }
+
+// Start implements node.Service, starting the data collection thread and the listening server of the dashboard.
+func (db *dashboard) Start(server *p2p.Server) error {
+	db.wg.Add(2)
+	go db.collectData()
+	go db.collectLogs() // In case of removing this line change 2 back to 1 in wg.Add.
+
+	http.HandleFunc("/", db.webHandler)
+	http.Handle("/api", websocket.Handler(db.apiHandler))
+
+	listener, err := net.Listen("tcp", fmt.Sprintf("%s:%d", db.config.Host, db.config.Port))
+	if err != nil {
+		return err
+	}
+	db.listener = listener
+
+	go http.Serve(listener, nil)
+
+	return nil
+}
+
+// Stop implements node.Service, stopping the data collection thread and the connection listener of the dashboard.
+func (db *dashboard) Stop() error {
+	var err error
+	// Close the connection listener
+	if err = db.listener.Close(); err != nil {
+		log.Warn("Failed to close listener", "err", err)
+	}
+
+	errc := make(chan error)
+	db.closing <- errc // collectData
+	<-errc
+	db.closing <- errc // collectLogs
+	<-errc
```

Comment:
Don't ignore errors, log at least, preferrably store and return at the end of the function.